pkg/arvo/app/chat-hook.hoon

```diff
@@ -51,8 +54,137 @@
+      =/  write=(unit permission)  (get-permission chat %write)
+      ?.  &(?=(^ read) ?=(^ write))
+        ~&  [%missing-permission chat read=?=(~ read) write=?=(~ write)]
+        [%white [(slav %p (snag 0 chat)) ~ ~]]
+      ::  village: exclusive to writers
+      ::
+      ?:  &(?=(%white kind.u.read) ?=(%white kind.u.write))
+        [%white who.u.write]
+      ::  channel: merge blacklists
+      ::
+      ?:  &(?=(%black kind.u.read) ?=(%black kind.u.write))
+        [%black (~(uni in who.u.read) who.u.write)]
+      ::  journal: exclusive to writers
+      ::
+      ?:  &(?=(%black kind.u.read) ?=(%white kind.u.write))
+        [%white who.u.write]
+      ::  mailbox: exclusive to readers
+      ::
+      ?:  &(?=(%white kind.u.read) ?=(%black kind.u.write))
+        [%white who.u.read]
+      ~|  [%weird-kinds kind.u.read kind.u.write]
+      !!
+    ::
+    ++  get-permission
+      |=  [chat=path what=?(%read %write)]
+      %^  scry:cc  (unit permission)
+        %permission-store
+      [%permission %chat (snoc chat what)]
+    ::
+    ++  make-poke
+      |=  [app=term =mark =vase]
+      ^-  card
+      [%pass /on-load/[app]/[mark] %agent [our.bol app] %poke mark vase]
+    ::
+    ++  delete-group
+      |=  [host=ship group=path]
+      ^-  card
+      ::  if we host the group, delete it directly
+      ::
+      ?:  =(our.bol host)
+        %^  make-poke  %group-store
+          %group-action
+        !>  ^-  group-action
+        [%unbundle group]
+      ::  else, just delete the sync in the hook
+      ::
+      %^  make-poke  %permission-hook
+        %permission-hook-action
+      !>  ^-  permission-hook-action
+      [%remove group]
+    ::
+    ++  create-group
+      |=  [group=path who=(set ship)]
+      ^-  (list card)
+      :~  %^  make-poke  %group-store
+            %group-action
+          !>  ^-  group-action
+          [%bundle group]
+        ::
+          %^  make-poke  %group-store
+            %group-action
+          !>  ^-  group-action
+          [%add who group]
+      ==
+    ::
+    ++  hookup-group
+      |=  [group=path =kind]
+      ^-  (list card)
+      :*  %^  make-poke  %permission-group-hook
+            %permission-group-hook-action
+          !>  ^-  permission-group-hook-action
+          [%associate group [group^kind ~ ~]]
+        ::
+          =/  =ship  (slav %p (snag 1 group))
+          ?.  =(our.bol ship)  ~
+          :_  ~
+          %^  make-poke  %permission-hook
+            %permission-hook-action
+          !>  ^-  permission-hook-action
+          [%add-owned group group]
+      ==
+    ::
+    ++  record-group
+      |=  [group=path chat=path]
+      ^-  card
+      =/  =metadata
+        ~|  [%weird-chat-path chat]
+        %*  .  *metadata
+          title         (snag 1 chat)
+          date-created  now.bol
+          creator       (slav %p (snag 0 chat))
+        ==
+      %^  make-poke  %metadata-store
+        %metadata-action
+      !>  ^-  metadata-action
```

Comment:
I had previously been bitten by `!>`ing without actually checking the type of the noun I was putting in, leading to crashes on the receiver side. Casting the contents of a `!>`, so you get to use the type system for sanity-checks. If the action type ever changes, we'll get a nest-fail instead of this silently breaking.

Reply:
:( I dislike it stylistically but understand your rationale and think it's reasonable. We need a shorthand for this